File: lib/java/src/main/java/io/github/msgpact/internal/generation/GenerateRandomUnion.java
package io.github.msgpact.internal.generation;

import static io.github.msgpact.internal.generation.GenerateRandomStruct.generateRandomStruct;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import io.github.msgpact.internal.types.VStruct;
import io.github.msgpact.internal.types.VTypeDeclaration;

public class GenerateRandomUnion {
    public static Object generateRandomUnion(Object blueprintValue,
            boolean useBlueprintValue, Map<String, VStruct> unionTagsReference,
            GenerateContext ctx) {
        if (useBlueprintValue) {
            final var startingUnion = (Map<String, Object>) blueprintValue;
            final var entry = startingUnion.entrySet().stream().findAny().get();
            final var unionTag = entry.getKey();
            final var unionStructType = unionTagsReference.get(unionTag);
            final var unionStartingStruct = (Map<String, Object>) startingUnion.get(unionTag);

            return Map.of(unionTag,
                    generateRandomStruct(unionStartingStruct, true, unionStructType.fields, ctx));
        } else {
            final var sortedUnionTagsReference = new ArrayList<>(unionTagsReference.entrySet());

            Collections.sort(sortedUnionTagsReference, (e1, e2) -> e1.getKey().compareTo(e2.getKey()));

            final var randomIndex = ctx.randomGenerator.nextIntWithCeiling(sortedUnionTagsReference.size());
            final var unionEntry = sortedUnionTagsReference.get(randomIndex);
            final var unionTag = unionEntry.getKey();
            final var unionData = unionEntry.getValue();

            return Map.of(unionTag,
                    generateRandomStruct(null, false, unionData.fields, ctx));
        }
    }
}

File: lib/go/internal/types/GenerateRandomUnion.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package types

import "sort"

// GenerateRandomUnion produces a pseudo-random union value respecting the supplied union tag definitions.
func GenerateRandomUnion(blueprintValue any, useBlueprintValue bool, unionTagsReference map[string]*TStruct, ctx *GenerateContext) map[string]any {
	if ctx == nil {
		return map[string]any{}
	}

	if !useBlueprintValue {
		sortedUnionTags := make([]string, 0, len(unionTagsReference))
		for key := range unionTagsReference {
			sortedUnionTags = append(sortedUnionTags, key)
		}
		sort.Strings(sortedUnionTags)

		if len(sortedUnionTags) == 0 {
			return map[string]any{}
		}

		randomIndex := 0
		if ctx.RandomGenerator != nil {
			randomIndex = ctx.RandomGenerator.NextIntWithCeiling(len(sortedUnionTags))
		}

		unionTag := sortedUnionTags[randomIndex]
		unionStruct := unionTagsReference[unionTag]
		structValue := GenerateRandomStruct(nil, false, unionStruct.Fields, ctx)
		return map[string]any{unionTag: structValue}
	}

	startingUnion := toStringAnyMap(blueprintValue)
	for unionTag, unionValue := range startingUnion {
		unionStruct := unionTagsReference[unionTag]
		startingStruct := toStringAnyMap(unionValue)
		structValue := GenerateRandomStruct(startingStruct, true, unionStruct.Fields, ctx)
		return map[string]any{unionTag: structValue}
	}

	return map[string]any{}
}
